packages/create-subscription/src/createSubscription.js

```diff
@@ -0,0 +1,156 @@
+/**
+ * Copyright (c) 2014-present, Facebook, Inc.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import React from 'react';
+import warning from 'fbjs/lib/invariant';
+
+export function createSubscription<
+  Property,
+  CreatedSubscription,
+  Value,
+>(config: {|
+  // Synchronously gets the value for the subscribed property.
+  // Return undefined if the subscribable value is undefined,
+  // Or does not support synchronous reading (e.g. native Promise).
+  +getValue: (source: Property) => Value,
+
+  // Setup a subscription for the subscribable value in props.
+  // Due to the variety of change event types, subscribers should provide their own handlers.
+  // Those handlers should not attempt to update state though;
+  // They should call the callback() instead when a subscription changes.
+  // You may optionally return a subscription value to later unsubscribe (e.g. event handler).
+  +subscribe: (
+    source: Property,
+    callback: (value: Value) => void,
+  ) => CreatedSubscription,
+
+  // Unsubsribe from the subscribable value in props.
+  // The subscription value returned from subscribe() is passed as the second parameter.
+  +unsubscribe: (source: Property, subscription: CreatedSubscription) => void,
+|}): React$ComponentType<{
+  children: (value: Value) => React$Element<any>,
+  source: any,
+}> {
+  const {getValue, subscribe, unsubscribe} = config;
+
+  warning(
+    typeof getValue === 'function',
+    'Subscription must specify a getValue function',
+  );
+  warning(
+    typeof subscribe === 'function',
+    'Subscription must specify a subscribe function',
+  );
+  warning(
+    typeof unsubscribe === 'function',
+    'Subscription must specify an unsubscribe function',
+  );
+
+  type Props = {
+    children: (value: Value) => React$Element<any>,
+    source: any,
+  };
+  type State = {
+    source: Property,
+    subscriptionWrapper: {
+      subscription?: CreatedSubscription,
+    },
+    value?: Value,
+  };
```

Comment:
Just curious why we define the types here rather than outside this method - it means they would be re-recreated every time this is called, but since it probably won't be called often not too worried.

Reply:
I think they have to be defined inline because of the `<>` parameters?